### DISCUSSION TITLE
Returning a Pillow image that only exists in memory

### DISCUSSION THREAD

**Discussion by Vuizur** [litestar-org/litestar#1010] [Q&A] [answered]
I have a program that programmatically creates an image using Pillow. I want to return it using Starlite. I tried combining a bit FastAPI-inspired code from Stackoverflow + following the docs a bit, but couldn't get it to work.

```python
@get(
    "randompic/{lang:str}",
    media_type='image/jpeg'
)
def randompic(lang: str) -> io.BytesIO:
    img = rdpgen.generate_random_definition_pic(lang)
    # return img as bytes
    buf = io.BytesIO()
    img.save(buf, format="JPEG")
    buf.seek(0)

    return Response(content=buf, media_type="image/jpeg")

```
If you send a request, then you get the response `{"status_code":500,"detail":"Unable to serialize response content"}`. What would be the best way to achieve this?

**Comment by Goldziher** [comment]
Have you tried using a streaming response: https://starlite-api.github.io/starlite/1.48/usage/5-responses/8-streaming-responses/

**Comment by provinzkraut** [comment, marked as answer]
Two things:

1. Your return annotation is incorrect. You are not returning `io.BytesIO` but a `Response`
2. If you want to send a "raw" response, you should provide bytes, not `io.BytesIO`.

Therefore, the fix looks like this:

```python
@get(
    "randompic/{lang:str}",
    media_type='image/jpeg'
)
def randompic(lang: str) -> Response[bytes]:
    img = rdpgen.generate_random_definition_pic(lang)
    # return img as bytes
    buf = io.BytesIO()
    img.save(buf, format="JPEG")
    buf.seek(0)

    return Response(content=buf.read(), media_type="image/jpeg")
```

Alternatively you could do what @Goldziher suggested and stream the file from memory, which depending on its size might be an advantage or a disadvantage.

**Reply by Vuizur** [reply]
Thank you a lot, this works flawlessly!
